Put the FreeIPA cluster behind a load balancer

jsf9k opened the pull request:
## 🗣 Description ##

This pull request puts the FreeIPA cluster behind a load balancer.

Please read the message associated with commit edb2f2525106cd766de372f195fa228bfc58db80.

See also:
- cisagov/ansible-role-freeipa-server#49
- cisagov/freeipa-server-tf-module/pull/65
- cisagov/freeipa-server-packer/pull/79

## 💭 Motivation and context ##

With these changes, if a FreeIPA server becomes unresponsive DNS will automatically be updated to remove the unresponsive server from the rotation.  It is hoped that these changes will help correct the errors described in cisagov/cool-system-internal#89.

## 🧪 Testing ##

All automated tests pass.  I currently have these changes deployed to our COOL staging environment, and they appear to be functioning as expected.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [ ] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] All new and existing tests pass.

jsf9k committed:
Add provider and remote state for public DNS

We will require this to create the necessary public DNS records for
ACM certificate validation.

jsf9k committed:
Add certificate and load balancers

jsf9k committed:
Add LB target groups

jsf9k committed:
Add LB listeners

jsf9k committed:
Add LB target group attachments

jsf9k committed:
Update security groups to work with load balancers

jsf9k committed:
Update dns module to work with LB

jsf9k committed:
Configure ALB target groups to send HTTP/2 requests to targets

jsf9k committed:
Move the ACM cert TF code to its own file

jsf9k committed:
Move the creation of the load balancers into a separate TF file

jsf9k committed:
Move creation of load balancer listeners into its own file

jsf9k committed:
Move the creation of the LB target groups into its own file

jsf9k committed:
Move creation of the LB target group attachments to its own file

jsf9k committed:
Simplify/coalesce load balancer target group attachments

jsf9k committed:
Add a comment to describe the calculation of the target type

jsf9k committed:
Refactor the three IPA server modules into a single map of modules

This allows us to simplify the LB target group attachments
considerably.

Note that this does leave the state in a slightly janky state, since
Terraform now wants to replace ipa1 and ipa2.  It wants to do this
because we didn't used to specify the realm input variable for those
two servers, since replicas never make use of that input variable.
Now that we are treating all servers identically, though, we specify
the realm for all three.  This causes a change to the instance
metadata that doesn't actually change anything, but Terraform detects
the change and wants to redeploy.  This jankiness will go away the
next time the FreeIPA instances are redeployed.

jsf9k committed:
Create the A records for the individual servers outside the dns module

The dns module should only create records for the IPA _cluster_.

jsf9k committed:
Remove a now-unused input variable

felddy commented on security_groups/client.tf:
```diff
@@ -16,3 +16,14 @@ resource "aws_security_group_rule" "client_egress" {
+  security_group_id = aws_security_group.client.id
+  type              = "egress"
+  protocol          = each.value.proto
+  cidr_blocks       = [for ip in var.load_balancer_ips : format("%s/32", ip)]
```
With regards to `load_balancer_ips`, is there a technical reason we're assuming the IPs are IPv4?  I'm not suggesting they shouldn't at this point.  I understand this is a pattern we've been following so far.  I'm trying to keep a picture of where we have assumptions in the system that might change.

jsf9k replied:
There is no technical reason.  Both ALBs and NLBs support running in dualstack mode.

jsf9k replied:
Can this conversation be resolved, or are you requesting something be changed?

dav3r commented on certificate.tf:
```diff
@@ -0,0 +1,46 @@
+
+  lifecycle {
+    create_before_destroy = true
+  }
```
Can you please add a comment here explaining why we want/need `create_before_destroy = true`?

jsf9k replied:
Please see commit d9fbb550aa611b1e20a00fb4baddd9e9e7a6290c.

dav3r replied:
Thanks!

dav3r commented on dns/README.md:
```diff
@@ -59,16 +52,20 @@ No modules.
-| ttl | The TTL value to use for Route53 DNS records (e.g. 86400).  A smaller value may be useful when the DNS records are changing often, for example when testing. | `number` | `86400` | no |
+| domain | The domain for the IPA cluster (e.g. example.com). | `string` | n/a | yes |
+| hostname | The hostname portion of the IPA FQDN (e.g. ipa).  This value and the value of the domain variable comprise the FQDN of the IPA cluster. | `string` | n/a | yes |
+| load\_balancer\_dns\_name | The Route53 DNS name of the Network Load Balancer in front of the IPA cluster (e.g. IPA-0123456789abcdef.elb.us-east-1.amazonaws.com). | `string` | n/a | yes |
+| load\_balancer\_zone\_id | The Route53 DNS zone ID of the Network Load Balancer in front of the IPA cluster (e.g. ZKX36JXQ8W93M). | `string` | n/a | yes |
```
Unnecessary capitalization?
```suggestion
| load\_balancer\_dns\_name | The Route53 DNS name of the network load balancer in front of the IPA cluster (e.g. IPA-0123456789abcdef.elb.us-east-1.amazonaws.com). | `string` | n/a | yes |
| load\_balancer\_zone\_id | The Route53 DNS zone ID of the network load balancer in front of the IPA cluster (e.g. ZKX36JXQ8W93M). | `string` | n/a | yes |
```

jsf9k replied:
I'm fine either way, but it's capitalized here because Network Load Balancer is the AWS-specific name of the resource type.  If you want me to change it that's fine, but please confirm.

dav3r replied:
That makes sense- it is fine as is.

dav3r commented on certificate.tf:
```diff
@@ -0,0 +1,46 @@
+  allow_overwrite = true
+  name            = each.value.name
+  records         = [each.value.record]
+  ttl             = 60
```
Should we be using `var.ttl` here?
```suggestion
  ttl             = var.ttl
```

jsf9k replied:
We should not.  These are just records that are put in place so that ACM can verify we own the domain.  If `var.ttl` is set to a large value because we don't expect the FreeIPA-specific DNS records to change, we would still want these records to have a low TTL so that redeployments can happen in a timely manner.

I can add a comment explaining this.

dav3r replied:
Got it- thanks, a comment will be helpful there.

jsf9k replied:
Please see commit ea461b50a2abe01cedaf12aa5524f22000007220.

dav3r commented on dns/variables.tf:
```diff
@@ -6,12 +6,22 @@
+
+variable "load_balancer_dns_name" {
+  type        = string
+  description = "The Route53 DNS name of the Network Load Balancer in front of the IPA cluster (e.g. IPA-0123456789abcdef.elb.us-east-1.amazonaws.com)."
+}
+
+variable "load_balancer_zone_id" {
+  type        = string
+  description = "The Route53 DNS zone ID of the Network Load Balancer in front of the IPA cluster (e.g. ZKX36JXQ8W93M)."
```
Unnecessary capitalization?
```suggestion
  description = "The Route53 DNS name of the network load balancer in front of the IPA cluster (e.g. IPA-0123456789abcdef.elb.us-east-1.amazonaws.com)."
}

variable "load_balancer_zone_id" {
  type        = string
  description = "The Route53 DNS zone ID of the network load balancer in front of the IPA cluster (e.g. ZKX36JXQ8W93M)."
```

jsf9k replied:
See [here](https://github.com/cisagov/cool-sharedservices-freeipa/pull/51#discussion_r817828809).

dav3r commented on dns/dns.tf:
```diff
@@ -1,120 +1,80 @@
-# Create all necessary DNS records for the IPA servers.
+# Create all necessary DNS records for the IPA cluster.
 # -------------------------------------------------------------------------------
 resource "aws_route53_record" "server_A" {
```
Should we make this corresponding name change (and update references to it)?
```suggestion
resource "aws_route53_record" "cluster_A" {
```

jsf9k replied:
Please see commit 3d05b3233c99761a61ea43cd98017ca3dea773e5.

dav3r commented on freeipa.tf:
```diff
@@ -24,22 +24,32 @@ module "security_groups" {
-module "ipa0" {
+module "ipa" {
   source = "github.com/cisagov/freeipa-server-tf-module"
+  # AWS reserves the first four and the last IP address in each
+  # subnet.
+  #
+  # cisagov/freeipa-server-tf-module now requires us to assign IPs in
+  # order to break the dependency of DNS record resources on the
+  # corresponding EC2 instance resources; otherwise, it is not
+  # possible to recreate the IPA servers one by one as is required
+  # when a new FreeIPA AMI is made available.
```
Did you mean to have this comment both above (in locals) and repeated here?

jsf9k replied:
Please see commit 66de4789f6817e69e68281d0f3868cac51dfa79e.

dav3r reviewed (approved):
Truly a tour de force here- this looks great!  I had a few questions and comments, but nothing that stands in the way of my approval.  💪 💼

jsf9k committed:
Temporarily use a non-default branch for freeipa-server-tf-module

This is necessary to test the changes made in
cisagov/freeipa-server-tf-module#65 and will be reverted once that PR
is merged.

jsf9k committed:
Add load_balancer_hostname input variable

jsf9k committed:
Add a comment explaining why we specify create_before_destroy=true

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Add a comment explaining why we use a hardcoded TTL value

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Remove duplicate comment

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Rename a resource

The new name better describes what the resource represents.

Also update the terraform-docs information in the README.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Update FreeIPA server module config for updated input variable names

jsf9k committed:
Add connection termination to all LB target groups

This feature terminates connections to a deregistered target after a
timeout, allowing connections to drain off in a cleaner fashion.

jsf9k committed:
Add missing rule to server security group

Servers must allow egress to the load balancer to kinit, etc.

jsf9k commented:
Unfortunately these changes didn't pan out.  Anything that uses GSSAPI cannot be behind a load balancer, since the principal must contain the actual hostname and not that of the load balancer.
